capa/features/extractors/ts/function.py

```diff
@@ -0,0 +1,131 @@
+import itertools
+from typing import Tuple, Iterator
+from dataclasses import dataclass
+
+from tree_sitter import Node
+
+import capa.features.extractors.ts.integer
+from capa.features.insn import API, Number, Property
+from capa.features.common import String, Feature
+from capa.features.address import Address
+from capa.features.extractors.ts.engine import TreeSitterExtractorEngine
+from capa.features.extractors.base_extractor import FunctionHandle
+
+PSEUDO_MAIN = "PSEUDO MAIN"
+
+
+@dataclass
+class TSFunctionInner:
+    node: Node
+    name: str
+    engine: TreeSitterExtractorEngine
+
+
+def is_pseudo_main_function(fh: FunctionHandle, engine: TreeSitterExtractorEngine) -> bool:
+    return (
+        fh.address == engine.get_default_address()
+        and fh.inner.node == engine.tree.root_node
+        and fh.inner.name == PSEUDO_MAIN
+    )
+
+
+def extract_strings(fn_node: Node, engine: TreeSitterExtractorEngine) -> Iterator[Tuple[Feature, Address]]:
+    for node, _ in engine.get_string_literals(fn_node):
+        yield String(engine.get_range(node).strip('"')), engine.get_address(node)
+
+
+def extract_integer_literals(fn_node: Node, engine: TreeSitterExtractorEngine) -> Iterator[Tuple[Feature, Address]]:
+    for node, _ in engine.get_integer_literals(fn_node):
+        parsed_int = capa.features.extractors.ts.integer.parse_integer(engine.get_range(node), engine.language)
+        if parsed_int is not None:
+            yield Number(parsed_int), engine.get_address(node)
+
+
+def get_imports(name: str, namespaces: set[str], engine: TreeSitterExtractorEngine) -> Iterator[str]:
+    for namespace in itertools.chain([""], namespaces):
+        joined_name = engine.language_toolkit.join_names(namespace, name)
+        if engine.language_toolkit.is_import(joined_name):
+            yield joined_name
+
+
+def get_properties(fn_node: Node, engine: TreeSitterExtractorEngine) -> Iterator[Tuple[Node, str]]:
+    for node, _ in engine.get_assigned_property_names(fn_node):
+        qualified_names = engine.language_toolkit.split_name(engine.get_range(node))
+        if len(qualified_names) > 1:
+            yield node, engine.language_toolkit.join_names(*qualified_names[1:])
+
+
+def get_classes(fn_node: Node, engine: TreeSitterExtractorEngine) -> set[str]:
+    return set(
+        name
+        for node, _ in engine.get_new_object_names(fn_node)
+        for name in get_imports(engine.get_range(node), engine.namespaces, engine)
+    )
+
+
+def extract_classes_(fn_node: Node, engine: TreeSitterExtractorEngine) -> Iterator[Tuple[Feature, Address]]:
```

Comment:
What does the underscore at the function name end indicate?

Reply:
This was supposed to be a single leading underscore as per https://peps.python.org/pep-0008/#public-and-internal-interfaces, fixed.